# STA130 Week 01 Homework 

Please see the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) for the list of topics covered in this homework assignment, and a list of topics that might appear during ChatBot conversations which are "out of scope" for the purposes of this homework assignment (and hence can be safely ignored if encountered)


<details class="details-example"><summary style="color:blue"><u>Introduction</u></summary>

### Introduction
    
A reasonable characterization of STA130 Homework is that it simply defines a weekly reading comprehension assignment. 
Indeed, STA130 Homework essentially boils down to completing various understanding confirmation exercises oriented around coding and writing tasks.
However, rather than reading a textbook, STA130 Homework is based on ChatBots so students can interactively follow up to clarify questions or confusion that they may still have regarding learning objective assignments.

> Communication is a fundamental skill underlying statistics and data science, so STA130 Homework based on ChatBots helps practice effective two-way communication as part of a "realistic" dialogue activity supporting underlying conceptual understanding building. 

It will likely become increasingly tempting to rely on ChatBots to "do the work for you". But when you find yourself frustrated with a ChatBots inability to give you the results you're looking for, this is a "hint" that you've become overreliant on the ChatBots. Your objective should not be to have ChatBots "do the work for you", but to use ChatBots to help you build your understanding so you can efficiently leverage ChatBots (and other resources) to help you work more efficiently.<br><br>

</details>

<details class="details-example"><summary style="color:blue"><u>Instructions</u></summary>

### Instructions
    
1. Code and write all your answers (for both the "Prelecture" and "Postlecture" HW) in a python notebook (in code and markdown cells) 
    
> It is *suggested but not mandatory* that you complete the "Prelecture" HW prior to the Monday LEC since (a) all HW is due at the same time; but, (b) completing some of the HW early will mean better readiness for LEC and less of a "procrastentation cruch" towards the end of the week...
    
2. Paste summaries of your ChatBot sessions (including link(s) to chat log histories if you're using ChatGPT) within your notebook
    
> Create summaries of your ChatBot sessions by using concluding prompts such as "Please provide a summary of our exchanges here so I can submit them as a record of our interactions as part of a homework assignment" or, "Please provide me with the final working verson of the code that we created together"
    
3. Save your python jupyter notebook in your own account and "repo" on [github.com](github.com) and submit a link to that notebook though Quercus for assignment marking<br><br>

</details>

<details class="details-example"><summary style="color:blue"><u>Prompt Engineering?</u></summary>

### Prompt Engineering?    
    
The questions (as copy-pasted prompts) are designed to initialize appropriate ChatBot conversations which can be explored in the manner of an interactive and dynamic textbook; but, it is nonetheless **strongly recommendated** that your rephrase the questions in a way that you find natural to ensure a clear understanding of the question. Given sensible prompts the represent a question well, the two primary challenges observed to arise from ChatBots are 

1. conversations going beyond the intended scope of the material addressed by the question; and, 
2. unrecoverable confusion as a result of sequential layers logial inquiry that cannot be resolved. 

In the case of the former (1), adding constraints specifying the limits of considerations of interest tends to be helpful; whereas, the latter (2) is often the result of initial prompting that leads to poor developments in navigating the material, which are likely just best resolve by a "hard reset" with a new initial approach to prompting.  Indeed, this is exactly the behavior [hardcoded into copilot](https://answers.microsoft.com/en-us/bing/forum/all/is-this-even-normal/0b6dcab3-7d6c-4373-8efe-d74158af3c00)...

</details>


### Marking Rubric (which may award partial credit) 

- [0.1 points]: All relevant ChatBot summaries [including link(s) to chat log histories if you're using ChatGPT] are reported within the notebook
- [0.2 points]: Reasonable well-written general definitions for Question "2.2"
- [0.3 points]: Demonstrated understanding regarding Question "4"
<!-- - [0.2 points]: A sensible justification for the choice in Question "7.4" -->
- [0.4 points]: Requested assessment of ChatBot versus google performance in Question "8.3"


## "Pre-lecture" HW [*completion prior to next LEC is suggested but not mandatory*]

#### 1. Pick one of the datasets from the ChatBot session(s) of the **TUT demo** (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> If your TA has not shared a relevant ChatBot session from their **TUT demo** through a piazza post and a Quercus announcement, the **TUT notebook** has links to example ChatBot sessions that you can use; or, ...
> 
> ```python
> # feel free to just use the following if you prefer...
> import pandas as pd
> url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
> df = pd.read_csv(url)
> df.isna().sum()
> ```
    
</details>

In [3]:
import pandas as pd

# URL of the dataset
url = "https://data.cityofnewyork.us/resource/vfnx-vebw.csv"

# Load the dataset
data = pd.read_csv(url)

# Display the first few rows of the dataset
print(data.head())

# Confirm that the dataset has missing values
data.isna().sum()

           x          y unique_squirrel_id hectare shift      date  \
0 -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1 -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2 -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3 -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4 -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   

   hectare_squirrel_number    age primary_fur_color highlight_fur_color  ...  \
0                        3    NaN               NaN                 NaN  ...   
1                        4    NaN               NaN                 NaN  ...   
2                        8    NaN              Gray                 NaN  ...   
3                       14  Adult              Gray                 NaN  ...   
4                        5  Adult              Gray            Cinnamon  ...   

    kuks  quaas  moans tail_flags tail_twitches  approaches  indifferent  \
0  False  False  False

x                               0
y                               0
unique_squirrel_id              0
hectare                         0
shift                           0
date                            0
hectare_squirrel_number         0
age                            54
primary_fur_color              15
highlight_fur_color           366
combination_of_primary_and      0
color_notes                   950
location                       20
above_ground_sighter           38
specific_location             835
running                         0
chasing                         0
climbing                        0
eating                          0
foraging                        0
other_activities              851
kuks                            0
quaas                           0
moans                           0
tail_flags                      0
tail_twitches                   0
approaches                      0
indifferent                     0
runs_from                       0
other_interact

Since the result of data.isna().sum() for the variables age, primary_fur_color, highlight_fur_color... is not 0, we confirmed that the dataset has missing values.

### ChatBot summaries
Since I selected a dataset from a ChatBot session from the TUT demo, the following summary of the ChatBot session is a link provided by our TA.

link: https://chatgpt.com/share/5965ee20-f003-4d7a-9f06-57409c080e1a

Here is a summary of our interaction:

Request for Amusing or Interesting Datasets:

You initially asked for amusing, funny, or otherwise interesting datasets, preferably in CSV format.
I provided links to several datasets from Kaggle, including "The Simpsons Episodes Dataset," "Spongebob Squarepants Scripts Dataset," "Pokemon with Stats," "Punny Headlines," and "Star Wars Characters." However, the links were not clickable.
Request for Clickable Links:

You requested clickable links for the datasets.
I provided clickable links for the previously mentioned datasets.
Request for Different Datasets:

You asked for other datasets different from the initial ones provided.
I provided new links to interesting datasets, including "Rick and Morty Characters," "Dog Breeds," "Harry Potter Characters," "Netflix Movies and TV Shows," and "Video Game Sales."
Provided Dataset Link:

You provided a link to a dataset from the City of New York and requested the code to load the data using pandas.
Pandas Code to Load Data:

I provided a Python code snippet using pandas to load the dataset from the given URL:
python
Copy code
import pandas as pd

URL of the dataset
url = "https://data.cityofnewyork.us/resource/vfnx-vebw.csv"

Load the dataset
data = pd.read_csv(url)

Display the first few rows of the dataset
print(data.head())
This summary encapsulates our interaction, highlighting the progression of requests and responses, including dataset suggestions and the code provided to load a specific dataset using pandas.

#### 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a `pandas` DataFrame has, and then

1. use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,  
2. write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> A good initial prompt to start would be be something like
> - "I've downloaded a dataset about characters from animal crossings (from https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv), and I'd like to know what columns of information I have and how much data I have"
> 
> You can further reduce the scope of your inquiry with if needed with something like
> - "I've already downloaded the data and want to understand the size (or dimensions) of the dataset to start with"
> 
> *Some ChatBots can upload your data and do this for you; but, extended usage of this feature [likely requires a paid subscription](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/GPT/SLS/00006_gpt3p5_LoadDataPaywall.md); and, anyway, you need to run the code yourself rather than having a ChatBot do that for you; and, for STA130 we don't want a ChatBot to just do the analysis for us; rather, we instead want ChatBots to help us understand the steps we need to take to analyze the data; so,* **you DO NOT need to purchase an upgraded version of any ChatBots**
> - Free-tier level ChatBots like [GPT4o-mini](https://chat.openai.com/) or [Copilot](https://copilot.microsoft.com/) (which is partially based on [ChatGPT4.0](https://chat.openai.com/), and which you have access to through your UofT account) are sufficiently sophisticated and perfectly appropriate for the STA130 course
    
</details>

In [4]:
num_rows = data.shape[0]
num_columns = data.shape[1]

print(f'The DataFrame has {num_rows} rows and {num_columns} columns.')

The DataFrame has 1000 rows and 31 columns.


Observations: In my dataset, each row is an observation, representing an entity or unit of analysis, such as a person, event, or transaction, for a total of 1,000 observations.
Variables: In my dataset, each column is a variable that describes some attribute of the observation, such as age, gender, or occupation, for a total of 31 variables. Variables can be numerical (quantitative) or categorical (qualitative), such as hectare_squirrel_number is numerical and gender is categorical.

### ChatBot summaries
Shown in Question 5.

#### 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Use your ChatBot session to help you create working examples of using  `df.describe()` and `df['column'].value_counts()` for your dataset (although note that the `.value_counts()` method is not really meant to be used for numeric variables, so if you dataset has only numeric variables, `.value_counts()` might not be particularly informative...)
>
> #### ChatBot Response Scope 
>     
> If prompts are not sufficiently focused you will likely get overly broad responses from the ChatBot, but you can always respond with subsequent refinement requests to appropriately limit the scope of the ChatBot responses to focus on addressing your actual content targets; so, 
> - an initially very general inquiry like, "I need help analyzing my data" will likely result in a ChatBot response suggesting a wide variety of approaches and techniques for summarizing your dataset; but, re-prompting the ChatBot with something like, "What's the simplest form of summarization of this dataset that I could do and how do I do it in Python?" or suggesting guidance using the specific summarization methods requested above will helpfully re-orient the ChatBot to your specific interests and needs
> 
> #### Jupyter Notebook Hints
> 
> Jupyter notebook printouts usaully don't show all of the data (when there's too much to show, like if `df.describe()` includes results for many columns), but the printouts just show enough of the data to give an idea of what the results are which is all we're looking for at the moment
> 
> - Consider dividing the code that ChatBot provides you into different jupyter notebook cells so that each cell concludes with a key printed result; the last line of code in a jupyter notebook cell will automatically print out in a formatted manner, so replacing something like `print(df.head())` with `df.head()` as the last line of a cell provides a sensible way to organize your code
> - The printout suggestions above are demonstrated in `STA130F24_CourseProject.ipynb` if looking at an example would be helpful to understand what they're getting at...
    
</details>

In [5]:
# General information about the DataFrame
info_summary = data.info()

# Descriptive statistics for numerical data
numerical_summary = data.describe()

# Descriptive statistics for categorical data
categorical_summary = data.describe(include='object')

# Display the summaries
print("General Information:")
print(info_summary)
print("\nNumerical Summary:")
print(numerical_summary)
print("\nCategorical Summary:")
print(categorical_summary)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 31 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   x                           1000 non-null   float64
 1   y                           1000 non-null   float64
 2   unique_squirrel_id          1000 non-null   object 
 3   hectare                     1000 non-null   object 
 4   shift                       1000 non-null   object 
 5   date                        1000 non-null   int64  
 6   hectare_squirrel_number     1000 non-null   int64  
 7   age                         946 non-null    object 
 8   primary_fur_color           985 non-null    object 
 9   highlight_fur_color         634 non-null    object 
 10  combination_of_primary_and  1000 non-null   object 
 11  color_notes                 50 non-null     object 
 12  location                    980 non-null    object 
 13  above_ground_sighter        962 no

### ChatBot summaries
Shown in Question 5.

#### 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by `df.shape` and what is reported by `df.describe()` with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> If the dataset you're using does not have (a) non-numeric variables and (b) missing values in numeric variables (e.g., the `"villagers.csv"` example above has only a single numeric variable `row_n` which has no missing values), instead download and use the [https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv](https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv)" data to answer this question  
>
> In (a) above, the "columns it analyzes" refers to the columns of the output of `df.describe()` which will only include "numeric" columns by default, but you can can see the names of all the columns in a dataset using `df.columns`; and, make sure `df.shape` is refering to the dataset you think it is... if you've loaded a different dataset it might not have been called `df`(!)
>
> **If you get any errors (for example related to column names), copy and paste them as a response to the ChatBot, and see if it can help you resove them by adding the suggested adjustments to your code and then reruning all your code to see if the changes have fixed the problem (and repeat this process as needed until the problems have been resolved).**
    
</details>

In [18]:
# Based on the answers to the previous question, we know that the dataset we are using has non-numeric variables and no missing values in numeric variables.

# Size of the dataset given by data.shape
A = data.shape

# Size of the dataset given by data.describe()
B = data.describe()

# Display the size of the dataset
print("Size of the dataset given by df.shape:")
print(A)
print("\nNumerical Summary:")
print(B)

Size of the dataset given by df.shape:
(1000, 31)

Numerical Summary:
                 x            y          date  hectare_squirrel_number
count  1000.000000  1000.000000  1.000000e+03              1000.000000
mean    -73.967020    40.781087  1.011853e+07                 4.170000
std       0.007698     0.010139  4.267014e+04                 3.151989
min     -73.980961    40.765008  1.006202e+07                 1.000000
25%     -73.972957    40.772389  1.007202e+07                 2.000000
50%     -73.968374    40.778225  1.012202e+07                 3.000000
75%     -73.959765    40.791285  1.014202e+07                 6.000000
max     -73.949722    40.800046  1.020202e+07                23.000000


The result of data.shape of our dataset has 1000 rows and 31 columns.
The result of data.describe() of our dataset show we have 4 numerical variables and each variable has 1000 non values.

Since we already noticed that the dataset we are using has non-numeric variables and no missing values in numeric variables. So, we expect the result in data.shape and data.describe() have different number of columns, and for data.shape to have the same number of rows as the count of data.describe()

(a) data.shape provides the total number of columns in the DataFrame, including both numerical and categorical data types. It gives a complete count of all variables in the dataset. While, data.describe() only analyzes numerical columns. Which are only relevant to numeric data. Since our dataset counts non-numeric variables, the number of columns it analyzes by data.describe() is less than the number of columns analyzes by data.shape.

(b) data.shape does not provide any information about the number of non-null entries in each column; it merely counts rows and columns. While, data.describe() "count" reported the number of non-null entries in each column analyzed. Since our dataset has no missing values in numeric variables, for data.shape to have the same number of rows as the count of data.describe().

### ChatBot summaries
Shown in Question 5.

#### 5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

- an "attribute", such as `df.shape` which does not end with `()`
- and a "method", such as `df.describe()` which does end with `()` 
   

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The fact that a "method" such as `df.describe()` ends with `()` suggests that "methods" are essentially something that we would call a "function" in programming language terminology; but, without getting too technical or "in the weeds", it might also be worth considering that we could also contrast what the difference is between a "function" in a programming language versus a "function" in mathematics...  
    
</details><br><br>

***Don't forget to ask for summaries of your ChatBot session(s) and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatGPT)!***<br><br>

The distinction between an "attribute" and a "method" relates to how they are defined and be used. The "attribute" holds state about the object, and does not perform any computation or behavior beyond storing and retrieving data. The "method" is a function associated with the object. It can perform tasks, execute calculations, and return values based on the object’s attributes.

In conclusion, I learned that the () is intended to help clarify whether you're simply retrieving information about the state of an object (attribute) or performing some operation or computation (method).

### ChatBot summaries
link: https://chatgpt.com/share/66e38f8d-a908-8006-947e-0f5ad7405c9d

Certainly! Here’s a summary of our exchanges today:

Dataset Dimensions: You uploaded a dataset and asked how to determine the number of rows and columns in a pandas DataFrame. I provided you with a Python code snippet using df.shape to calculate and print these dimensions. The DataFrame was found to have 1,000 rows and 31 columns.

Definitions of Observations and Variables:

Observations: Defined as individual entries or records in the dataset, each representing a unique unit or entity, like a character in your dataset.
Variables: Defined as attributes or characteristics recorded for each observation, quantifiable into numerical or categorical data.
Summary of Dataset Columns:

I outlined methods to provide summaries of columns in the DataFrame using pandas functions like df.describe(), df.info(), and df.describe(include='object'), which help understand numerical and categorical data.
Discrepancies in Data Summaries:

I explained why there might be discrepancies between the output of df.shape and df.describe() concerning the number of columns analyzed and the count of non-null entries, highlighting the default behavior of df.describe() to focus only on numerical data unless specified otherwise.
Attributes vs. Methods in Python:

Discussed the distinction between attributes and methods in Python, noting that attributes (accessed without parentheses) represent data associated with an object, while methods (called with parentheses) are functions that perform operations or calculations on an object’s data.
This summary encapsulates the guidance provided on managing and analyzing your dataset with pandas, along with clarifying some Python programming concepts. You can use this overview to prepare your homework assignment record. If you need further details or adjustments to this summary, feel free to ask!

<details class="details-example"><summary style="color:blue"><u>Continue now...?</u></summary>

### Prelecture VS Postlecture HW
    
Feel free to work on the "Postlecture" HW below if you're making good progress and want to continue: in this case this is particularly reasonable as questions "6" and "7" below directly follow up and extend the "Prelecture" HW questions

*The benefits of continue would are that (a) it might be fun to try to tackle the challenge of working through some problems without additional preparation or guidance; and (b) this is a very valable skill to be comfortable with; and (c) it will let you build experience interacting with ChatBots (and beginning to understand their strengths and limitations in this regard)... it's good to have sense of when using a ChatBot is the best way to figure something out, or if another approach (such as course provided resources or a plain old websearch for the right resourse) would be more effective*
    
</details>    


## "Post-lecture" HW [*submission along with "Pre-lecture" HW is due prior to next TUT*]

#### 6. The `df.describe()` method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The answers here actually make it obvious why these can only be calculated for numeric variables in a dataset, which should help explain the answer to "4(a)" and "4(b)" above
>   
> Also notice that when `df.describe()` is used missing values are not explicitly removed, but `df.describe()`  provides answers anyway. Is it clear what `df.describe()` does with the data in each columns it analyzes if there is missing data in the column in question? 
>
> The next questions addresses removing rows or columns from a dataset in order to explicitly remove the presense of any missingness in the dataset (assuming we're not going to fill in any missing data values using any missing data imputation methods, which are beyond the scope of STA130); so, the behavior of `df.describe()` hints that explicitly removing missing may not always be necessary; but, the concern, though, is that not all methods may be able to handle missing data the way `df.describe()` does...
    
</details>

In [20]:
#Just have a peek on how the result of data.describe() looks like

# Descriptive statistics for numerical data
numerical_summary = data.describe()

# Display the summaries
print("\nNumerical Summary:")
print(numerical_summary)


Numerical Summary:
                 x            y          date  hectare_squirrel_number
count  1000.000000  1000.000000  1.000000e+03              1000.000000
mean    -73.967020    40.781087  1.011853e+07                 4.170000
std       0.007698     0.010139  4.267014e+04                 3.151989
min     -73.980961    40.765008  1.006202e+07                 1.000000
25%     -73.972957    40.772389  1.007202e+07                 2.000000
50%     -73.968374    40.778225  1.012202e+07                 3.000000
75%     -73.959765    40.791285  1.014202e+07                 6.000000
max     -73.949722    40.800046  1.020202e+07                23.000000


Count: This shows the number of non-null entries in each column (variable).

Mean: This calculate the average value of the data in each column (variable) by summing up all the values and dividing by the count of values.

Std: Standard Deviation, which measures the amount of variation or dispersion of a set of values. A low standard deviation means that the values tend to be close to the mean, whereas a high standard deviation means the values are spread out over a wider range.

Min: This shows the smallest value in each column (variable).

25%: Lower quartile, this indicates that 25% of the data values in each column (variable) are below this value.

50%: Median value, this value separates the higher half of a data sample from the lower half. For an even number of observations, it is the average of the two middle numbers.

75%: Upper quartile, this indicates that 75% of the data values in each column (variable) are below this value.

Max: This shows the Largest value in each column (variable).

### ChatBot summaries
Just check the definition again with ChatBot to get a more accurate definition.

link: https://chatgpt.com/share/66e39852-34cc-8006-b510-5007a4e8bd80

The df.describe() method in pandas provides a summary of the descriptive statistics for each numerical column in a DataFrame. Here's what each statistic represents:

Count: The number of non-null entries in the column. This statistic shows how many values are present, excluding any missing values.

Mean: The average of the values in the column. It is calculated by summing all the values and dividing by the count of values.

Std (Standard Deviation): A measure of the amount of variation or dispersion in a set of values. A low standard deviation means that the values tend to be close to the mean, whereas a high standard deviation means that the values are spread out over a wider range.

Min (Minimum): The smallest value in the column.

25% (25th Percentile): The value below which 25% of the data falls. It is a measure of the lower quartile.

50% (Median or 50th Percentile): The middle value of the data set when it is ordered from lowest to highest. This is also known as the median, which divides the data into two equal parts.

75% (75th Percentile): The value below which 75% of the data falls. It is a measure of the upper quartile.

Max (Maximum): The largest value in the column.

These statistics are helpful for getting a quick understanding of the distribution and central tendencies of the data, as well as its range and variability.

#### 7. Missing data can be considered "across rows" or "down columns".  Consider how `df.dropna()` or `del df['col']` should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

1. Provide an example of a "use case" in which using `df.dropna()` might be peferred over using `del df['col']`<br><br>
    
2. Provide an example of "the opposite use case" in which using `del df['col']` might be preferred over using `df.dropna()` <br><br>
    
3. Discuss why applying `del df['col']` before `df.dropna()` when both are used together could be important<br><br>
    
4. Remove all missing data from one of the datasets you're considering using some combination of `del df['col']` and/or `df.dropna()` and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.<br><br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Start a new ChatBot session **[but remember to first ask your ChatBot for summaries of your current session and perhaps coding results (so you can supply these in the homework as requested)]**, since your last ChatBot session has likely gotten quite long and has covered a lot of material at this point 
> - It can sometimes be helpful to reset ChatBot sessions to refocus them on the topics of inquiry without too much backlog history that might unintentionally bias things in certain directions and, of course, you can always re-introduce material from earlier conversations as it's relevant, such as for answering "D" based on reintroducing and updating code you made in a previous ChatBot session.  
> 
> #### ChatBot Scope Guidance
> 
> - This question is not interested in the general benefits of imputing missing data, or the general benefits of using `df.dropna()` and/or `del df['col']` to remove missing data, just how to most efficiently remove missing data if a user chooses to do so
> 
> - More sophisticated analyses for "filling in" rather than removing missing data (as considered here) are possible (based on making assumptions about missing data and using specific imputation methods or models) but these are "beyond the scope" of this homework assignment so this topics can be safely ignored for now
> 
> #### ChatBot Code Troubleshooting
> 
> A key issue to be aware of when asking ChatBots for help with something is that they are not running and checking code for correctess, and they often intertwine written instructions with code instructions; so, BEFORE YOU RUN ANY CODE provided by a ChatBot, you should check the following
> 
> 1. If this code changes an object or data, are you sure you want to run this code?
> 2. Can you easily "undo" the results of running code (e.g., from a copy `df_saved=df.copy()` or reloading the data) if running the code doesn't do what you want?
> 3. Is the state of the data what is expected by the code? Or have the objects been updated and changed so they're no longer what the code expects them to be? 
> 
> #### If you get any `Python` errors, copy and paste them into the ChatBot prompt and see if it can help you resove them; but, keep in mind the final point above becasue the ChatBot might not be aware of the state of your objects relative to the code it's producing...

</details><br>

1. When you want to analyze a dataset where each columns are crucial to the analysis and you cannot afford to lose any of them, but some rows have missing values.

    Example: Suppose you have a dataset of student ratings for each aspect of a course, where each column represents the rating for one aspect of the course. To better draw recommendations from this dataset, if a few rows have missing data for one or two columns but the rest of the data is intact, you may want to remove only those rows to keep the rest of the dataset intact.
    
2. When a column has a high percentage of missing values, rendering it unreliable or unusable for the analysis.

    Example: Consider the same dataset as above, with one column collecting students' ratings group presentations for learning the course better, but this course does not have group presentations, so most students are missing data in this column. In this case, missing data percentage is very high, it more practical to remove this column entirely.
    
3. If df.dropna() is applied first, you may end up dropping a significant number of rows unnecessarily due to missing values in a column that you would eventually delete.

In [24]:
# Confirm how many missing values that the dataset has
data.isna().sum()

x                               0
y                               0
unique_squirrel_id              0
hectare                         0
shift                           0
date                            0
hectare_squirrel_number         0
age                            54
primary_fur_color              15
highlight_fur_color           366
combination_of_primary_and      0
color_notes                   950
location                       20
above_ground_sighter           38
specific_location             835
running                         0
chasing                         0
climbing                        0
eating                          0
foraging                        0
other_activities              851
kuks                            0
quaas                           0
moans                           0
tail_flags                      0
tail_twitches                   0
approaches                      0
indifferent                     0
runs_from                       0
other_interact

4. As the reasoning I given in Answer7.3, the variables highlight_fur_color, color_notes, specific_location, other_activities and other_interactions have really high percentage of missing data, which render it unusable for my analysis. So I would apply del df['col'] to drop this 5 columns first, then apply df.dropna() to remove rows with missing data in the remaining columns to investigate the impact of each variable.

In [30]:
# Delete columns with high percentage of missing data
df = data.copy()
del df['highlight_fur_color']
del df['color_notes']
del df['specific_location']
del df['other_activities']
del df['other_interactions']

#Delete rows with missing data in the rest columns
df.dropna(inplace=True)

# Confirm that the dataset has no missing values
df.isna().sum()

x                             0
y                             0
unique_squirrel_id            0
hectare                       0
shift                         0
date                          0
hectare_squirrel_number       0
age                           0
primary_fur_color             0
combination_of_primary_and    0
location                      0
above_ground_sighter          0
running                       0
chasing                       0
climbing                      0
eating                        0
foraging                      0
kuks                          0
quaas                         0
moans                         0
tail_flags                    0
tail_twitches                 0
approaches                    0
indifferent                   0
runs_from                     0
geocoded_column               0
dtype: int64

After our approach of modifying the data with combination of del df['col'] and/or df.dropna(), we remove all missing values in the dataset.
### ChatBot summaries
link: https://chatgpt.com/share/66e3a4f8-8ef0-8006-8213-565d5d0906c0

Certainly! Here's a summary of our exchanges today, structured to capture the key elements of each discussion:

1. Handling Missing Data in DataFrames
We discussed various methods for handling missing data in pandas DataFrames, focusing on when to use df.dropna() versus del df['col'].

Use Case for df.dropna():

Scenario: When rows have missing values in critical measurements or information, but all columns are essential.
Example: Clinical trial results where complete columns are needed, but some rows may be dropped if they have missing values.
Use Case for del df['col']:

Scenario: When a column has a high percentage of missing data, making it unreliable.
Example: Customer feedback dataset where the 'feedback_score' column is mostly missing.
Importance of Operation Order:

Discussed the importance of removing columns with high missing data before dropping rows. This helps in minimizing data loss and maintaining the integrity of crucial data.
2. Practical Example of Data Cleaning
You presented a scenario involving a dataset with columns like 'Patient_ID', 'Age', 'BMI', 'Blood_Pressure', 'Cholesterol_Level', and 'Smoking_Status'. We discussed an approach to clean this dataset:

Before Cleaning:

Total rows: 1000
Columns with missing data, especially 'Cholesterol_Level' having 70% missing.
Approach:

Step 1: Remove 'Cholesterol_Level' due to its high missing data percentage.
Step 2: Apply df.dropna(inplace=True) to clean remaining rows with missing data.
After Cleaning:

Total rows reduced to 800, with complete data in all remaining columns.
3. Code Adjustments for DataFrame Operations
You intended to remove multiple columns using the del statement. I provided the correct method to achieve this, recommending the use of .drop() for multiple columns and explaining the proper use of del for single columns. Additionally, I emphasized the need to set inplace=True in df.dropna() to ensure changes persist.

Each part of our discussion aimed to enhance your understanding of data handling techniques in pandas, focusing on efficient and effective ways to manage and clean data sets. This summary encapsulates the key points and steps discussed, tailored for submission as a record of our interactions for your homework assignment.





    
#### 8. Give brief explanations in your own words for any requested answers to the questions below

> This problem will guide you through exploring how to use a ChatBot to troubleshoot code using the "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" data set 
> 
> To initialially constrain the scope of the reponses from your ChatBot, start a new ChatBot session with the following slight variation on the initial prompting approach from "2" above
> - "I am going to do some initial simple summary analyses on the titanic data set I've downloaded (https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv) which has some missing values, and I'd like to get your help understanding the code I'm using and the analysis it's performing"
        
1. Use your ChatBot session to understand what `df.groupby("col1")["col2"].describe()` does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you
    
> If needed, you can help guide the ChatBot by showing it the code you've used to download the data **AND provide it with the names of the columns** using either a summary of the data with `df.describe()` or just `df.columns` as demonstrated [here](../CHATLOG/COP/00017_copilot_groupby.md)
    
2. Assuming you've not yet removed missing values in the manner of question "7" above, `df.describe()` would have different values in the `count` value for different data columns depending on the missingness present in the original data.  Why do these capture something fundamentally different from the values in the `count` that result from doing something like `df.groupby("col1")["col2"].describe()`?

> Questions "4" and "6" above address how missing values are handled by `df.describe()` (which is reflected in the `count` output of this method); but, `count` in conjunction with `group_by` has another primary function that's more important than addressing missing values (although missing data could still play a role here).

3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT<br><br>
    
    1. Forget to include `import pandas as pd` in your code 
       <br> 
       Use Kernel->Restart from the notebook menu to restart the jupyter notebook session unload imported libraries and start over so you can create this error
       <br><br>
       When python has an error, it sometimes provides a lot of "stack trace" output, but that's not usually very important for troubleshooting. For this problem for example, all you need to share with ChatGPT or search on google is `"NameError: name 'pd' is not defined"`<br><br>

    2. Mistype "titanic.csv" as "titanics.csv"
       <br> 
       If ChatBot troubleshooting is based on downloading the file, just replace the whole url with "titanics.csv" and try to troubleshoot the subsequent `FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'` (assuming the file is indeed not present)
       <br><br>
       Explore introducing typos into a couple other parts of the url and note the slightly different errors this produces<br><br>
      
    3. Try to use a dataframe before it's been assigned into the variable
       <br> 
       You can simulate this by just misnaming the variable. For example, if you should write `df.groupby("col1")["col2"].describe()` based on how you loaded the data, then instead write `DF.groupby("col1")["col2"].describe()`
       <br><br>
       Make sure you've fixed your file name so that's not the error any more<br><br>
        
    4. Forget one of the parentheses somewhere the code
       <br>
       For example, if the code should be `pd.read_csv(url)` the change it to `pd.read_csv(url`<br><br>
        
    5. Mistype one of the names of the chained functions with the code 
       <br>
       For example, try something like `df.group_by("col1")["col2"].describe()` and `df.groupby("col1")["col2"].describle()`<br><br>
        
    6. Use a column name that's not in your data for the `groupby` and column selection 
       <br>
       For example, try capitalizing the columns for example replacing "sex" with "Sex" in `titanic_df.groupby("sex")["age"].describe()`, and then instead introducing the same error of "age"<br><br>
        
    7. Forget to put the column name as a string in quotes for the `groupby` and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question
       <br>
       For example, something like `titanic_df.groupby(sex)["age"].describe()`, and then `titanic_df.groupby("sex")[age].describe()`
        


In [10]:
#8
# URL of the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
df1 = pd.read_csv(url)

num_rows = df1.shape[0]
num_columns = df1.shape[1]

print(f'The DataFrame has {num_rows} rows and {num_columns} columns.')

The DataFrame has 891 rows and 15 columns.


In [35]:
#8.1
df1.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [16]:
#8.1
df1.groupby("survived")["pclass"].describe()

count      mean       std  min  25%  50%  75%  max
survived                                                    
0         549.0  2.531876  0.735805  1.0  2.0  3.0  3.0  3.0
1         342.0  1.950292  0.863321  1.0  1.0  2.0  3.0  3.0

It breaks down our titanic dataset into groups based on the unique entries in "col1" which is 0 and 1, 1 is survived and 0 is not survived. Then calculates several statistical measures for "col2" which is the class of the passangers within each of these groups. These measures include count, mean, standard deviation, minimum, 25% percentile, median, 75% percentile and maximum.

From the results we can see that there were 342 surviving passengers and 549 dead passengers, and the cabin class of the surviving passengers was on average lower than that of the dead passengers.

#### Q8.2
In df.describe(), the count shows the number of non-null entries in each column across the entire dataset. While, df.groupby("col1")["col2"].describe() shows the number of non-null entries for "col2" within each category of "col1".

This is fundamentally different because it reveals how the missing data is distributed across different categories or groups within the dataset. This is particularly useful for determining whether data in certain categories are systematically missing.

In [1]:
# URL of the dataset
url = "https://data.cityofnewyork.us/resource/vfnx-vebw.csv"

# Load the dataset
data = pd.read_csv(url)

# Display the first few rows of the dataset
print(data.head())

# Confirm that the dataset has missing values
data.isna().sum()

NameError: name 'pd' is not defined

In [4]:
# URL of the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"

# Load the dataset
df1 = pd.read_csv(url)

num_rows = df1.shape[0]
num_columns = df1.shape[1]

print(f'The DataFrame has {num_rows} rows and {num_columns} columns.')

HTTPError: HTTP Error 404: Not Found

In [5]:
DF1.groupby("survived")["pclass"].describe()

NameError: name 'DF1' is not defined

In [6]:
df1 = pd.read_csv(url

SyntaxError: incomplete input (3101177560.py, line 1)

In [11]:
df1.group_by("survived")["pclass"].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'

In [12]:
df1.groupby("Survived")["pclass"].describe()

KeyError: 'Survived'

In [13]:
df1.groupby("survived")["Pclass"].describe()

KeyError: 'Column not found: Pclass'

In [14]:
df1.groupby(survived)["pclass"].describe()

NameError: name 'survived' is not defined

In [15]:
df1.groupby("survived")[pclass].describe()

NameError: name 'pclass' is not defined

#### 8.3
A. For this error, it's easier to work in a ChatBot session to fix the errors. When I share this error in the ChatBot session, ChatGPT fix the coding errors immediately. The google search does not provides the necessary toubleshooting help more quickly than ChatGPT.

B. Since, I use the url instead of downloaded file. It shows HTTPError: HTTP Error 404: Not Found instead of FileNotFoundError: [Errno 2]. For this error, it's easier to work in a ChatBot session to fix the errors. When I share this error in the ChatBot session, ChatGPT point out the typo error and provide me the correct address immediately. The google search didn't provide any troubleshooting help regard the typos in the first few results.

C. For this error, it's easier to work in a ChatBot session to fix the errors. When I share this error in the ChatBot session, ChatGPT point out the variable naming and possible case-sensitivity issues immediately. The google search only show variable naming issues, does not provide a possibility of where potential errors may occur.

D. For this error, it's easier to work in a ChatBot session to fix the errors. Both ChatGPT and google search effectively pointed out the problem and gave the solution. But ChatGPT is more convenient because it directly provides you with the correct modified code.

E. Same. For this error, it's easier to work in a ChatBot session to fix the errors. Both ChatGPT and google search effectively pointed out the problem and gave the solution. But ChatGPT is more convenient because it directly provides you with the correct modified code.

F. The first error is KeyError: 'Survived', the second error is KeyError: 'Column not found: Pclass'. For this error, it's easier to work in a ChatBot session to fix the errors. When I share this error in the ChatBot session, ChatGPT provides a step-by-step improvement process, such as the checking code for column names, and possible errors in existing code. The google search does not provide such a detailed solution.

G. Both encounter the NameError: name '' is not defined. For this error, it's easier to work in a ChatBot session to fix the errors. Both ChatGPT and google search effectively pointed out the problem and gave the solution. But ChatGPT is more comprehensive and convenient, because it directly provides you with the correct modified code and point out the potential issues that column name is used directly in a function or condition without being correctly quoted as a string.

### ChatBot summaries
link: https://chatgpt.com/share/66e3b537-600c-8006-9e98-0f058fd3d578

Here’s a summary of our exchanges regarding your homework assignment involving the Titanic dataset:

Loading the Dataset: You attempted to load the Titanic dataset from a URL but faced a connectivity issue initially. After resolving that, you corrected the URL to properly load the dataset using pandas.

Understanding GroupBy and Describe: I explained the use of df.groupby("col1")["col2"].describe() in pandas, which groups data by col1 and provides descriptive statistics for col2 within those groups.

Handling Errors in Python Code:

NameError and SyntaxError: You encountered several errors due to misnaming variables and functions (Df1 vs. df1, group_by vs. groupby) and syntax issues (unclosed parentheses). I guided you on correcting these errors by ensuring proper case sensitivity and syntax in your Python code.
KeyError: You faced issues with incorrect column names due to case sensitivity (Survived vs. survived, Pclass vs. pclass). I advised you to check column names using df.columns and to ensure correct spelling and casing in your code.
HTTPError: You encountered a 404 Not Found error due to using a mistyped URL. After correcting the URL, you were able to access the dataset.
Coding Best Practices: I provided guidance on best practices for coding in Python with pandas, emphasizing the importance of case sensitivity, using string literals for column names, and proper syntax to avoid common errors like NameError and SyntaxError.

This record encapsulates our troubleshooting process and the solutions provided for the issues you encountered while working with the Titanic dataset for your homework assignment. If you need further details or additional information, feel free to ask!

#### 9. Have you reviewed the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?<br>
    
<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Just answering "Yes" or "No" or "Somewhat" or "Mostly" or whatever here is fine as this question isn't a part of the rubric; but, the midterm and final exams may ask questions that are based on the tutorial and lecture materials; and, your own skills will be limited by your familiarity with these materials (which will determine your ability to actually do actual things effectively with these skills... like the course project...)
    
</details>
    
***Don't forget to ask for summaries of your ChatBot session(s) and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatGPT)!***

# Recommended Additional Useful Activities [Optional]

The "Ethical Profesionalism Considerations" and "Current Course Project Capability Level" sections below **are not a part of the required homework assignment**; rather, they are regular weekly guides covering (a) relevant considerations regarding professional and ethical conduct, and (b) the analysis steps for the STA130 course project that are feasible at the current stage of the course<br><br>

<details class="details-example"><summary style="color:blue"><u>Ethical Professionalism Considerations</u></summary>

## Ethical Professionalism Considerations

> If the observed data is "no events occured" does this mean the data is "missing" and [should be ignored](https://priceonomics.com/the-space-shuttle-challenger-explosion-and-the-o)?
> 
> - NASA: \<determines temperature doesn't affects "o-ring" by subseting data to just "o-ring" incidents\>
> - Also NASA: \<launches the shuttle on a cold day\>

|No apparent "o-ring" failure and temperature relationship|Apparent between "o-ring" failure and temperature relationship|
|:-|:-|
if you just look at "o-ring" failure event data|if you instead look at ALL the data as you should|
|![](https://etzq49yfnmd.exactdn.com/wp-content/uploads/2022/03/image06-14.png)|![](https://etzq49yfnmd.exactdn.com/wp-content/uploads/2022/03/image02-33.png)|
|![](https://upload.wikimedia.org/wikipedia/commons/8/8b/Shuttle_Challenger_explosion.gif?20190203170223)|![](https://i.makeagif.com/media/10-04-2014/nT57xW.gif)|

<br>
    
</details>    

<details class="details-example"><summary style="color:blue"><u>Current Course Project Capability Level</u></summary>

## Current Course Project Capability Level

> The data we'll use for the STA130 course project is based on the [Canadian Social Connection Survey](https://casch.org/cscs). Please see the [data use agreement](https://static1.squarespace.com/static/60283c2e174c122f8ebe0f39/t/6239c284d610f76fed5a2e69/1647952517436/Data+Use+Agreement+for+the+Canadian+Social+Connection+Survey.pdf) regarding the appropriate and ethical professional use of this data (available at the bottom of the [CSCS](https://casch.org/cscs) webpage).
> 
> 1. Have a very quick look at the list of available variables using the [link](https://drive.google.com/file/d/1ISVymGn-WR1lcRs4psIym2N3or5onNBi/view) (again at the bottom of the [CSCS](https://casch.org/cscs) webpage); then, 
> 2. examine the code in the first thirteen code cells of [STA130F24_CourseProject.ipynb](https://github.com/pointOfive/stat130chat130/blob/main/CP/STA130F24_CourseProject.ipynb) to get an initital understanding of how we might subset to different studies included in the [data](https://drive.google.com/file/d/1mbUQlMTrNYA7Ly5eImVRBn16Ehy9Lggo/view) (again accessible at the bottom of the [CSCS](https://casch.org/cscs) webpage); then,     
> 3. review the fourteenth and fifteenth cells (with the comments "Here's a high level summary of the data" and "And here are some explanations about the columns in the data") a little more closely to get a better sense of which columns seem to be the most interesting and whether or not they seem to have a lot of missing data
    
</details>        

### Afterward

Here are few ideas of some other kinds of interactions you might consider exploring with a ChatBot...

> While these are likely to be extremely practically valuable, they are not a part of the homework assignment, so do not include anything related to these in your homework submission

- With respect to improving ones ability in statistics, coding, communication, and other key data science skills
    - what is the ChatBots perception its own capabilities and uses as an AI-driven assistance tool 
    - and does ChatBots assessment of itself influence or agree with your own evalution of the ChatBot? 

- ChatBots can introduce and explain the "World War 2 planes" problem and the "Monte Hall" problem... 
    - how well does do they seem to do and introducing and explaining other "unintuitive surprising statistics paradoxes"?

- If you consider the process of writing about why you chose to take this course, and the skills you were hoping to build through this course with respect to your current ideas about what possible careers 
    - and how do you think the exercise would be different if you framed it as a dialogue with a ChatBot
    - and do you think the difference could be positive and productive, or potentially biasing and distracting?
    
- ChatBots sometimes immediately responds in simple helpful ways, but other times it gives a lot of extraneous information that can be overwheling... are you able to prompt and interact with ChatBots in manner that keeps its reponses helpful and focused on what you're interested in? 

- ChatBots tends to respond in a fairly empathetic and supportive tone...
    - do you find it helpful to discuss concerns you might have about succeeding in the course (or entering university more generally) with a ChatBot?
    
- For what purposes and in what contexts do you think a ChatBot could provide suggestions or feedback about your experiences that might be useful? 
